core/src/main/java/org/apache/iceberg/BaseOverwriteFiles.java

```diff
@@ -127,8 +140,20 @@ protected void validate(TableMetadata base) {
       }
     }
 
-    if (conflictDetectionFilter != null && base.currentSnapshot() != null) {
-      validateAddedDataFiles(base, startingSnapshotId, conflictDetectionFilter, caseSensitive);
+    if (appendConflictDetectionFilter != null && base.currentSnapshot() != null) {
+      validateAddedDataFiles(base, startingSnapshotId, appendConflictDetectionFilter, caseSensitive);
+    }
+
+    boolean validateNewDeletes = deleteConflictDetectionFilter != null && base.currentSnapshot() != null;
```

Comment:
I think that the behavior here should be slightly different. There are two concerns: 1) whether to check delete files for snapshot isolation and 2) what conflict detection filter to use. Basing `validateNewDeletes` on whether the conflict detection filter was set doesn't seem correct to me.

I don't think there is a case where we don't want to validate delete files if we have called `validateFromSnapshot` to set the base snapshot. I think that we should add this as a boolean field that is set when `validateFromSnapshot` is called.

Then, if we are validating delete files, we should have two separate checks. First, if there are any files in `deletedDataFiles`, then we perform the validation below. If the conflict detection filter wasn't set, then we should use `Expressions.alwaysTrue` to find candidate delete files. Second, if an overwrite filter was set, then we should run `validateNoNewDeletes` with either the delete filter or the delete conflict detection filter. The conflict detection filter should be an optimization, not a way to turn off delete validations.

I think that makes the API more understandable and consistent.

Reply:
Here's what I changed this to locally while thinking through it:

```java
    // validateDeletes is set to true in validateFromSnapshot. Maybe we should default it if that method isn't called?
    if (validateDeletes) {
      if (deletedDataFiles.size() > 0) {
        validateNoNewDeletesForDataFiles(
            base, startingSnapshotId, deleteConflictDetectionFilter,
            deletedDataFiles, caseSensitive);
      }

      if (rowFilter() != Expressions.alwaysFalse()) {
        if (deleteConflictDetectionFilter != null) {
          validateNoNewDeletes(base, startingSnapshotId, deleteConflictDetectionFilter, caseSensitive);
        } else {
          validateNoNewDeletes(base, startingSnapshotId, rowFilter(), caseSensitive);
        }
      }
    }
```

Reply:
> Basing `validateNewDeletes` on whether the conflict detection filter was set doesn't seem correct to me.

If I got you correctly, you are proposing that `validateFromSnapshot` will now indicate whether we should validate delete files. I think that is different compared to how `RowDelta` and `OverwriteFiles` work right now. I'd actually say calling `validateFromSnapshot` is an optimization that tells us from which snapshot to start looking. We never validate new appends if the append conflict detection filter is null. Moreover, it is not always possible to set the starting snapshot. If we start on an empty table, we must validate all snapshots. Here is our copy-on-write commit logic.

```
Long scanSnapshotId = scan.snapshotId();
if (scanSnapshotId != null) {
  overwriteFiles.validateFromSnapshot(scanSnapshotId);
}

Expression conflictDetectionFilter = conflictDetectionFilter();
overwriteFiles.validateNoConflictingAppends(conflictDetectionFilter);
``` 

Also, in your snippet, why call `validateNoNewDeletesForDataFiles` if we already know the overwrite filter is set? I think `validateNoNewDeletesForDataFiles` is simply a more efficient version of `validateNoNewDeletes` that can open delete files that match the filter to check their content for conflicts. The problem is that we can use `validateNoNewDeletesForDataFiles` only if we overwrite specific files.